### ISSUE TITLE
Warn user with regression loss if they have an activation function on their final layer

### ISSUE THREAD

**Issue by EdwardDixon** [PyTorchLightning/pytorch-lightning#4489]
## 🚀 Detect accidental use of activation function when training a model on a regression problem
We train a lot of classifiers, so it is easy to forget that when we train a model on a regression problem, we should _not_ use an activation function on the output layer.  Even worse, the model will "sort of" learn, and it can take a long, long time to figure out what the problem is ([see here, Twitter thread between me, your Official PyTorch Lightning Account, and Eminent Machine Learning Personage](https://twitter.com/EdwardDixon3/status/1321851291560300544)).

### Motivation

Although is easy to mistakenly use that final activation layer, it is hard to spot the error.  However, it should be very detectable programmatically.

### Pitch

Although it's hard to figure out why your model trains so badly in this scenario, in principle it does seem very amenable to a programmatic check by the Trainer:

- Is a regression loss is being used (https://pytorch.org/docs/stable/generated/torch.nn.MSELoss.html)?
- If so, check _final_ layer if model to see if an activation function is being used
- If activation function present, warn user and link to something appropriate [like this Stack Overflow article](https://stackoverflow.com/questions/46680481/activation-function-for-output-layer-for-regression-models-in-neural-networks/46688108)

### Alternatives

A day or two of swearing, following by a deep sense of embarrassment.

**Comment by s-rog** [comment]
There are legitimate uses for activation functions for regression outputs though, like using relu(6) to clamp your output.

Obvious bugs often get the best of us :]
